src/annotator/draw-tool.tsx

```diff
@@ -0,0 +1,230 @@
+function normalizeRect(r: Rect): Rect {
+  const minX = Math.min(r.left, r.right);
+  const maxX = Math.max(r.left, r.right);
+  const minY = Math.min(r.top, r.bottom);
+  const maxY = Math.max(r.top, r.bottom);
+  return {
+    type: 'rect',
+    left: minX,
+    top: minY,
+    right: maxX,
+    bottom: maxY,
+  };
+}
+
+/**
+ * Errors while drawing a shape using {@link DrawTool}.
+ */
+export class DrawError extends Error {
+  constructor(message = 'Drawing failed') {
+    super(message);
+  }
+}
+
+/** Specifies the type of shape to draw. */
+export type Tool = 'rect' | 'point';
+
+/**
+ * Tool for drawing shapes for use as the target region of an annotation.
+ *
+ * When drawing is active, DrawTool creates an overlay into which the incomplete
+ * shape is drawn. The shape is updated in response to user gestures.
+ *
+ * Drawing is initiated using {@link DrawTool.draw}. Only one shape can be
+ * drawn at a time.
+ */
+export class DrawTool implements Destroyable {
+  private _container: HTMLElement;
+
+  /** Surface for current drawing. */
+  private _surface?: SVGSVGElement;
+
+  private _tool: Tool;
+
+  /** Current drawing shape. */
+  private _shape?: Rect;
+
+  /** Callback for when draw operation ends successfully. */
+  private _drawEnd?: (s: Shape) => void;
+
+  /** Callback for when draw operation ends with an error. */
+  private _drawError?: (e: DrawError) => void;
+
+  /** Controller for ending the drawing operation. */
+  private _abortDraw?: AbortController;
+
+  /**
+   * @param root - Container in which the user can draw a shape
+   */
+  constructor(root: HTMLElement) {
+    this._container = root;
+    this._tool = 'rect';
+    if (!isPositioned(root)) {
+      root.style.position = 'relative';
+    }
+  }
+
+  destroy() {
+    this.cancel();
+  }
+
+  /**
+   * Begin drawing a shape.
+   *
+   * @param tool - Type of shape to draw
+   * @return - Promise for the shape drawn by the user
+   */
+  async draw(tool: Tool): Promise<Shape> {
+    this._tool = tool;
+
+    // Only one drawing operation can be in progress at a time.
+    this.cancel();
+
+    // Create a transparent SVG canvas overlaid on top of the container, with
+    // a crosshair cursor to indicate the user can click to draw.
+    const surface = document.createElementNS(
+      'http://www.w3.org/2000/svg',
+      'svg',
+    );
+    surface.setAttribute('data-testid', 'surface');
+    surface.setAttribute('width', '100%');
+    surface.setAttribute('height', '100%');
+    surface.style.cursor = 'crosshair';
+    surface.style.position = 'absolute';
+    surface.style.left = '0px';
+    surface.style.top = '0px';
+    surface.style.zIndex = '10';
+    this._container.append(surface);
+    this._surface = surface;
+
+    let resolve: (s: Shape) => void;
+
+    const shape = new Promise<Shape>((resolve_, reject_) => {
+      resolve = resolve_;
+      this._drawEnd = resolve_;
+      this._drawError = reject_;
+    });
```

Comment:
I reckon you can use `promiseWithResolvers()` here.

```suggestion
    const { promise: shape, resolve, reject } = promiseWithResolvers<Shape>();
    this._drawEnd = resolve;
    this._drawError = reject;
```